### PR TITLE
Fix S2583 FP: Delegate unsubscription

### LINKED ISSUE #8094: Fix S2583/S2589 FP: Delegate can be null when last invocation is removed but comparing the delegate against null gets reported as always false
### Description

When removing the last invocation in a delegate, it can become null (see https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/subtraction-operator#delegate-removal), but any comparison after that will be flagged as unnecessary because Sonar still identifies the delegate as non null.
It is correctly identified as potentially null by VS analyzers, but Sonar still thinks it can't be null. See this screenshot as reference:
image

### Repro steps

```
    internal class Test
    {
        public void TestMethod() 
        {
            Action? someDelegate = delegate { };
            someDelegate += Callback;
            
            // ....

            someDelegate -= Callback;

            if (someDelegate == null) // S2583 reported here
            {
                // ...
            }
            
        }

        private void Callback()
        {
            //...
        } 
    }
```

### Expected behavior

Sonar correctly identifies the delegate as potentially null and don't issue any warning.

### Actual behavior

Sonar incorrectly identifies the delegate as not null and issues a S2583 warning. It could also be a S2589 depending on the code inside the if.

### Related information

* C#/VB.NET Plugins version = 7.3.0.77872
* Visual Studio version = 17.7.2
* MSBuild / dotnet version =7.0
* Operating System = Windows 11

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Fixes #8094

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs [hunk 1]
```diff
@@ -51,6 +52,29 @@ private static ProgramState ProcessNumericalCompoundAssignment(ProgramState stat
...
+        // For this reason, we remove the NotNull constraint from the delegate instance.
+        if (assignment.Target.Type.TypeKind == TypeKind.Delegate
+            && assignment.OperatorKind == BinaryOperatorKind.Subtract
+            && state.HasConstraint(assignment.Target, ObjectConstraint.NotNull))
+        {
+            var value = (state[assignment.Target] ?? SymbolicValue.Empty).WithoutConstraint(ObjectConstraint.NotNull);
+            state = state
+                .SetOperationValue(assignment, value)
+                .SetOperationValue(assignment.Target, value);
+            state = state.SetSymbolValue(assignment.Target.TrackedSymbol(state), value);
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs:69]: Note: the left side of the -= operator must be a field, property, or indexer. In the first two cases, the symbol is always available. For indexers, I haven't been able to create a test case where the indexer would have `NotNull` as a constraint.
e.g.
```cs
this[42] = delegate { };
this[42] += Callback; // this[42] doesn't have the NotNull constraint
this[42] -= Callback; 
```
So I don't check whether the tracked symbol is `null`.

Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs:69]: Is `Target` always a `TrackedSymbol`? What about indexers?

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]
Looks very nice! Clean implementation. Just add a few more UTs where the delegate is not a local variable (field, proper, indexer).

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs (hunk 1 above)
Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs]: AFAIK the engine doesn't store Null constraints for indexers, so it's not going to get inside the if statement. I've added a UT to cover it.

**Pull request merged**